### ISSUE TITLE
Problems getting metrics from Azure in Grafana

### ISSUE THREAD

**Issue by Marco-Schmitz-KPS** [grafana/grafana#42937]
Hi there,

within Azure there is the service principal with the roles "Monitoring Reader" and "Reader" assigned.

Therefor the resource group "elkprodresourcegroup" within the subscription "IT Commerce PROD" is accessable.

This resource group has 3017 records, at least 6 of type "SQL database" and 5 of type "SQL server" which are of interest right now.

Using Grafana 7.5.5 or Grafana 8.3.1 together with Azure Monitor as data source leads to the same error behavior.

So I choose:
- Service: "Metrics"
- Subscription: "IT Commerce PROD"
- Resource group: "elkprodresourcegroup"
- Namespace (only Grafana 7.5.5): "Microsoft.Cache/Redis"
- Resource type (only Grafana 8.3.1): "Azure Cache for Redis"

From my perspective it seems entirely wrong that Namespace or Resource type only shows "Azure Cache for Redis".

I'd expect that more types are shown here, like all the types of the 3017 records, for example "Storage account", "SQL database", "SQL server", "Kubernetes service", "Key vault"...

![image2021-12-9_11-48-19](https://user-images.githubusercontent.com/67626973/145415323-c82b7527-58cb-452b-ae11-dec72325a0c4.png)

Using commandline after logging in with that service principal I am able to get some results for:
`az monitor metrics list-namespaces --resource /subscriptions/2ee14c5f-****-****-****-************/resourceGroups/elkprodresourcegroup/providers/Microsoft.Sql/servers/elkproddbsrvprocessdatadeltaprod -o json`

This prooves that the service principal has enough rights to access this.

![image2021-12-9_13-22-53](https://user-images.githubusercontent.com/67626973/145415356-700e427a-aa21-4ad3-b0f7-e7fbb1218041.png)

I just want within Grafana to be able to work on the metrics of the SQL database "elkproddbprocessdatadeltaprod".

The web ui shows metrics like e.g. "CPU percentage":

Best regards,
Marco Schmitz

**Comment by Marco-Schmitz-KPS** [comment]
Hello there,

so it took me some time to get used to the Azure REST API.

Right now I am successfully requesting “SQL database (not working in Grafana)” as well as “Azure Cache for Redis (the only one working in Grafana)” from Azure REST API.

But it only works if I use different api-version parameters.

For example for “SQL database (not working in Grafana)” I need “api-version=**2021-08-01-preview**”.
https://management.azure.com/subscriptions/2ee14c5f-****-****-****-************/resourceGroups/elkprodresourcegroup/providers/Microsoft.Sql/servers/elkproddbsrvprocessdatadeltaprod/databases/elkproddbprocessdatadeltaprod?api-version=2021-08-01-preview

But for “Azure Cache for Redis (the only one working in Grafana)” I need “api-version=**2021-06-01**”.
https://management.azure.com/subscriptions/2ee14c5f-****-****-****-************/resourceGroups/elkprodresourcegroup/providers/Microsoft.Cache/Redis/elkprodrediscachedeltaprod?api-version=2021-06-01

Using some wrong api-version like “api-version=**2021-04-01**” I get an error with an response containing JSON naming the api-versions I have to use.


So I started to think if Grafana is using the wrong api-version, one that fits for “Azure Cache for Redis (the only one working in Grafana)” but not for “SQL database (not working in Grafana)”.

But on https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts#L25 I found:
```
apiVersion = '2018-01-01';
apiPreviewVersion = '2017-12-01-preview';
```
Using these version I get the error with that response containing JSON for both resources.

So right now I am confused how Grafana solves this?

Can I configure the api-version within Azure for some specific resources like the SQL database in order to gain more compatibility ?


All the best,
Marco Schmitz

**Comment by sunker** [comment]
Thanks for detailed issue @Marco-Schmitz-KPS! 

>Can I configure the api-version within Azure for some specific resources like the SQL database in order to gain more compatibility

Currently it's not possible to change api version in Grafana. If different resource types in Azure Monitor are supposed to use different api versions, it's going to require lots of changes on our side. Would like to get some input from Microsoft on this. @kostrse or @mitulk can you chime in?

**Comment by Marco-Schmitz-KPS** [comment]
Hi @sunker,
In parallel I am in contact with Azure support. And I have to correct myself, because the apiVersion was a dead end. To be honest I was using the wrong URLs, so here comes the update.

The structure for a GET request regarding the metrics is: 
`https://management.azure.com/{{resourceUri}}/providers/Microsoft.Insights/metrics?api-version=2018-01-01 `

SQL database (not working in Grafana): 
`https://management.azure.com/subscriptions/2ee14c5f-****-****-****-************/resourceGroups/elkprodresourcegroup/providers/Microsoft.Sql/servers/elkproddbsrvprocessdatadeltaprod/databases/elkproddbprocessdatadeltaprod/providers/Microsoft.Insights/metrics?api-version=2018-01-01`

Azure Cache for Redis (the only one working in Grafana):
`https://management.azure.com/subscriptions/2ee14c5f-****-****-****-************/resourceGroups/elkprodresourcegroup/providers/Microsoft.Cache/Redis/elkprodrediscachedeltaprod/providers/Microsoft.Insights/metrics?api-version=2018-01-01
`
Using Postman I was able to receive metrics from both resources.

And as you can see, the apiVersion is the same as it is used in the Grafana sourcecode.

But something is still causing these problems so that I cannot get my metrics...

**Comment by andresmgot** [comment]
Hi @Marco-Schmitz-KPS,

I am not able to reproduce the issue with my account. As you can see, the SQL database can be queried:

![Screenshot from 2021-12-23 15-18-28](https://user-images.githubusercontent.com/4025665/147252786-ce5778b6-98c0-4b38-b575-694ced657c66.png)

For some reason, your database is not being detected. Please go to browser console and check the API response when you select a resource group, we are using that response to populate the rest of selectors. It should be something like: 

```
/api/datasources/<ds_id>/resources/azuremonitor/subscriptions/<sub_iod>/resourceGroups/elkprodresourcegroup/resources?api-version=2018-01-01
```

What's the response there? Can you locate the SQL database in there?

**Comment by Marco-Schmitz-KPS** [comment]
Hi @andresmgot,
so this is my request:

`GET | http://localhost:3000/api/datasources/1/resources/azuremonitor/subscriptions/2ee14c5f-****-****-****-************/resourceGroups/elkprodresourcegroup/resources?api-version=2018-01-01
`

The response contains 999 JSON Objects and a nextLink.

Since this resource group has 3017 records I think there is kind of a paging going on here.

Within the first 999 JSON Objects there is no "SQL".

**Comment by sunker** [comment]
Okay - yes it looks like this is bug in Grafana. I'll add this to our backlog. Thanks for detailed issue @Marco-Schmitz-KPS!

**Comment by andresmgot** [comment]
Got it, this will be fixed as a side effect of https://github.com/grafana/grafana/pull/43522 since we are now filtering by resource type, you will only get SQL databases (will be available in Grafana 8.4.0). In any case, we should handle the case in which there are more than one thousand records for a resource type.
